File: infrastructure/db-core/src/main/java/com/nmh/commerce/coupon/FixedPriceDiscountPolicyEntity.java
package com.nmh.commerce.coupon;

import com.nmh.commerce.domain.Money;
import com.nmh.commerce.utils.MoneyConverter;
import jakarta.persistence.Convert;
import jakarta.persistence.DiscriminatorValue;
import jakarta.persistence.Entity;
import lombok.AccessLevel;
import lombok.NoArgsConstructor;

@Entity
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@DiscriminatorValue("FixedPriceDiscountPolicy")
public class FixedPriceDiscountPolicyEntity extends DiscountPolicyEntity {

    @Convert(converter = MoneyConverter.class)
    private Money discountPrice;

    public FixedPriceDiscountPolicyEntity(CouponEntity coupon, Money discountPrice) {
        super(coupon);
        this.discountPrice = discountPrice;
    }

    public static FixedPriceDiscountPolicyEntity from(FixedPriceDiscountPolicy fixedPriceDiscountPolicy, CouponEntity coupon) {
        return new FixedPriceDiscountPolicyEntity(coupon, fixedPriceDiscountPolicy.discountPrice);
    }

    @Override
    public DiscountPolicy toDomain() {
        return new FixedPriceDiscountPolicy(discountPrice);
    }
}

File: infrastructure/db-core/src/main/kotlin/com/nmh/commerce/coupon/FixedPriceDiscountPolicyEntity.kt
package com.nmh.commerce.coupon

import com.nmh.commerce.domain.Money
import com.nmh.commerce.utils.MoneyConverter
import jakarta.persistence.Convert
import jakarta.persistence.DiscriminatorValue
import jakarta.persistence.Entity

@Entity
@DiscriminatorValue("FixedPriceDiscountPolicy")
class FixedPriceDiscountPolicyEntity(
    coupon: CouponEntity,
    @field:Convert(converter = MoneyConverter::class)
    private val discountPrice: Money,
) : DiscountPolicyEntity(coupon = coupon) {
    override fun toDomain(): DiscountPolicy = FixedPriceDiscountPolicy(discountPrice)

    companion object {
        fun from(
            fixedPriceDiscountPolicy: FixedPriceDiscountPolicy,
            coupon: CouponEntity,
        ): FixedPriceDiscountPolicyEntity = FixedPriceDiscountPolicyEntity(coupon, fixedPriceDiscountPolicy.discountPrice)
    }
}
